allennlp/training/trainer.py

```diff
@@ -24,288 +19,66 @@
-            An integer specifying the CUDA device to use. If -1, the CPU is used.
-        grad_norm : ``float``, optional, (default = None).
-            If provided, gradient norms will be rescaled to have a maximum of this value.
-        grad_clipping : ``float``, optional (default = ``None``).
-            If provided, gradients will be clipped `during the backward pass` to have an (absolute)
-            maximum of this value.  If you are getting ``NaNs`` in your gradients during training
-            that are not solved by using ``grad_norm``, you may need this.
-        learning_rate_scheduler : ``PytorchLRScheduler``, optional, (default = None)
-            A Pytorch learning rate scheduler. The learning rate will be decayed with respect to
-            this schedule at the end of each epoch. If you use
-            :class:`torch.optim.lr_scheduler.ReduceLROnPlateau`, this will use the ``validation_metric``
-            provided to determine if learning has plateaued.  To support updating the learning
-            rate on every batch, this can optionally implement ``step_batch(batch_num_total)`` which
-            updates the learning rate given the batch number.
-        summary_interval: ``int``, optional, (default = 100)
-            Number of batches between logging scalars to tensorboard
-        histogram_interval : ``int``, optional, (default = ``None``)
-            If not None, then log histograms to tensorboard every ``histogram_interval`` batches.
-            When this parameter is specified, the following additional logging is enabled:
-                * Histograms of model parameters
-                * The ratio of parameter update norm to parameter norm
-                * Histogram of layer activations
-            We log histograms of the parameters returned by
-            ``model.get_parameters_for_histogram_tensorboard_logging``.
-            The layer activations are logged for any modules in the ``Model`` that have
-            the attribute ``should_log_activations`` set to ``True``.  Logging
-            histograms requires a number of GPU-CPU copies during training and is typically
-            slow, so we recommend logging histograms relatively infrequently.
-            Note: only Modules that return tensors, tuples of tensors or dicts
-            with tensors as values currently support activation logging.
-        should_log_parameter_statistics : ``bool``, optional, (default = True)
-            Whether to send parameter statistics (mean and standard deviation
-            of parameters and gradients) to tensorboard.
-        should_log_learning_rate : ``bool``, optional, (default = False)
-            Whether to send parameter specific learning rate to tensorboard.
-        log_batch_size_period : ``int``, optional, (default = ``None``)
-            If defined, how often to log the average batch size.
-        """
+                 serialization_dir: str,
```

Comment:
like, maybe this is even too restrictive.

for example, if you wanted to train a GAN you'd have _two_ `Model`s, right?

Reply:
Well, yeah, though I think it's fine to focus on single-model options for now.  I wouldn't try to make our abstraction fit GANs or other things until we've actually tried to build them ourselves - it's hard to get the abstraction right if you don't fully understand the needs of the use case.

This is a lot better than what was there, and once we have more experience with other training regimes, we can improve it again (which might just involve having a `SingleModelTrainer` base class, though I'm wary of making class hierarchies too deep).